### ISSUE TITLE
ICMP DNAT not work

### ISSUE THREAD

**Issue by cmheia** [ajaybhargav/lwip_nat#1]
It seems that [`iphdr->dest.addr = nat_entry.cmn->source.addr;`](https://github.com/ajaybhargav/lwip_nat/blob/c0ab071a24d4266de6da268566bc5e5f6498704f/src/core/ipv4/ip4_nat.c#L565) failed to get valid ip address since [`IPNAT_ENTRY_RESET(nat_entry.cmn);`](https://github.com/ajaybhargav/lwip_nat/blob/c0ab071a24d4266de6da268566bc5e5f6498704f/src/core/ipv4/ip4_nat.c#L523) already reset it, so [`err = in_if->output(in_if, q, (ip_addr_t *)&(iphdr->dest));`](https://github.com/RT-Thread/rt-thread/blob/3691b67e83ab57196dfc62606b1be18db76994a7/components/net/lwip_nat/ipv4_nat.c#L574) will generate a broadcast.

**Comment by ajaybhargav** [comment]
IPNAT_ENTRY_RESET only sets the TTL value to 0 does not touch anything else. Are you facing any issue or is it just a doubt?

**Comment by cmheia** [comment]
Emm, it's my mistake.
My version of `IPNAT_ENTRY_RESET` is:
```c
#define IPNAT_ENTRY_RESET(x) do { \
    ip_addr_set_any(&(x)->src); \
    ip_addr_set_any(&(x)->dest); \
    (x)->ttl = 0; \
} while(0)
```
This NAT imp not tracing TCP state, and always use same port for mapping([`LWIP_NAT_DEFAULT_STATE_TABLES_TCP`](https://github.com/ajaybhargav/lwip_nat/blob/c0ab071a24d4266de6da268566bc5e5f6498704f/src/core/ipv4/ip4_nat.c#L905) + [`LWIP_NAT_DEFAULT_TCP_SOURCE_PORT`](https://github.com/ajaybhargav/lwip_nat/blob/c0ab071a24d4266de6da268566bc5e5f6498704f/src/core/ipv4/ip4_nat.c#L925)).

I'm try to prevent new connection be mixed up with that already shutdown but still have `ttl != 0` by [`ip4_nat_tcp_lookup_incoming`](https://github.com/ajaybhargav/lwip_nat/blob/c0ab071a24d4266de6da268566bc5e5f6498704f/src/core/ipv4/ip4_nat.c#L866) with modify `IPNAT_ENTRY_RESET` which cause this.

**Comment by ajaybhargav** [comment]
> This NAT imp not tracing TCP state, and always use same port for mapping

Agree... Can you share your improvements so I can merge them in the source?

**Comment by cmheia** [comment]
These code is in company's repo and I can't copy them out.
But my solution is simpily add a state machine similar to `TCP/IP ILLustrated Volume 1: The Protocols` Figure 18-12(Client side).
```c
static
void ip4_nat_state_tracer_step (ip4_nat_entry_tcp_t     *entry,
                                const struct tcp_hdr     *tcphdr,
                                enum nat_state_trace_dir  dir)
```
And call this in `ip4_nat_udp_lookup_incoming` and `ip4_nat_udp_lookup_outgoing`.
Finally, if a TCP connection is switching to `TIME_WAIT`, do this:
```c
sys_timeout(1000, ip4_nat_purge_closed_tcp, (void *)nat_entry);
```

```c
static
void ip4_nat_purge_closed_tcp (ip4_nat_entries_tcp_t *nat_entry)
{
    IPNAT_ENTRY_RESET(&nat_entry->cmn);
}
```

**Comment by ajaybhargav** [comment]
> These code is in company's repo and I can't copy them out.

Aww sad. Its GPL though, Maybe your boss is ready to share it to make this world a happy place 😄
Anyways... Thanks for the pointers.
